Delay LaTeX feature tests as long as possible

orlitzky opened the pull request:
We shouldn't need a functional LaTeX engine to turn the matrix `[1]` into `\left(\begin{array}{r}1\end{array}\right)`. We used to try; now we don't. (This avoids compiling a test document in a temporary directory.)

There's some other cleanup at the same time, but the basic idea is to avoid putting the result `default_engine()` in the preferences dict when the preferences dict is constructed. Instead, we now leave the "engine" option set to `None` by default. Inside the functions where latex is actually used, we check...

1. Did the user pass `engine=whatever` to this function?
2. If not, is the "engine" option set in the preferences dict?
3. If not, use the default engine.

So `foo(engine=bar)` overrides the preferences dict, and the preferences dict overrides `default_engine()`.

Closes:

* https://github.com/sagemath/sage/issues/39351

orlitzky committed:
src/sage/misc/latex.py: drop the "engine_name" option

This does nothing, except waste your time trying to ensure that it is
consistent with the "engine" property and that It iS PrOpErLy
CapItALizEd.

orlitzky committed:
src/sage/misc/latex.py: default the "engine" option to None

This delays probing the default engine (which tries to compile some
LaTeX code in a temporary directory) when you merely wish to query
e.g. the matrix delimiters. A later commit will re-insert the probe
at a later point.

The doctest for the options dict was updated to print a sorted list of
items, so now its output is no longer random.

orlitzky committed:
src/sage/misc/latex.py: invoke default_engine() as late as possible

When we try to find the current LaTeX engine, the preferences dict
will now have __options["engine"] == None to indicate that we should
fall back to the default engine. In each case we check for None, and
invoke default_engine() only if necessary.

This allows us to query the options dict without triggering the
expensive is_functional() method for the LaTeX features.

orlitzky committed:
src/sage/misc/latex.py: add a test for the laziness of default_engine()

We should not need to test the functionality of various engines before
we can e.g. obtain the latex representation of a matrix. We add a
doctest to ensure that this is the case.

Closes GH-39351

user202729 commented on src/sage/misc/latex.py:
```diff
@@ -485,16 +485,35 @@ def default_engine():
     from sage.features.latex import pdflatex, xelatex, lualatex
     if lualatex().is_present():
-        return 'lualatex', 'LuaLaTeX'
+        return 'lualatex'
```
Looks like dangerous API change to me. Is this function public API? (it doesn't even start with an underscore?)

orlitzky replied:
It's one year old, not in the default namespace and not used outside of `sage.misc.latex`. Obviously you're right, but I bet I could give it the underscore it always wanted and no one would notice.

orlitzky replied:
I added the underscore

orlitzky replied:
@kwankyu are you OK with me making `default_engine()` an internal/underscore method in `sage.misc.latex`?

kwankyu replied:
No.  I think converting a public function to an internal function is an even bigger API change. Though "no one would notice", I see no reason to treat it as an exception to the deprecation principle.

On the other hand, I don't remember why the latex option list contains "engine_name", which is used nowhere within Sage. I agree that it is better to drop it as it is not really an option.

I suggest to leave `default_engine()` in public API, but change its behavior dropping the engine name. If you do so, please follow the proper deprecation procedure. If it is cumbersome, how about leaving `default_engine()` as it is, but just dropping "engine_name" from the latex option list?

orlitzky replied:
I guess I will put the old `default_engine` (no underscore) back, and deprecate it. Then we will have `_default_engine` with no "engine_name", internal from now on, but without breaking the API for anyone.

kwankyu replied:
OK.

orlitzky committed:
src/sage/misc/latex.py: make default_engine() an internal function

This function is relatively new and is only used in sage.misc.latex.
I don't think there is a compelling need to have it as part of the public
API, so let's give it an underscore.

kwankyu commented on src/sage/misc/latex.py:
```diff
@@ -474,27 +474,46 @@ def has_latex_attr(x) -> bool:
+def _default_engine():
+    r"""
     Return the default latex engine and the official name of the engine.
 
```
The docstring should change.

orlitzky replied:
Thanks, done.

orlitzky committed:
src/sage/misc/latex.py: bring back default_engine() for deprecation

In case anyone is using this, we avoid removing it (or changing its
return type) by surprise, and instead deprecate it.

changes. Lines starting # with '#' will be ignored, and an empty
message aborts the commit.  # # On branch lazier-latex-feature-tests #
Changes to be committed: # modified: src/sage/misc/latex.py #

orlitzky committed:
src/sage/misc/latex.py: update docstring for _default_engine()

This function no longer returns the stylized engine name, so its
docstring needs an update to reflect that.

kwankyu commented on src/sage/misc/latex.py:
```diff
@@ -496,6 +502,48 @@ def default_engine():
+        \end{array}\right)
+        sage: sage.misc.latex._default_engine = real_de
+
+    """
```
No empty line at the end, L536.

kwankyu replied:
Check also a linter complaint in the same file.

orlitzky replied:
both done

orlitzky committed:
src/sage/misc/latex.py: move a newline somewhere else

(This appeases pycodestyle, and makes the docstring endings more
consistent.)

kwankyu reviewed (approved):
Thanks. LGTM.
